src/main/java/org/apache/sysds/runtime/matrix/data/LibCommonsMath.java

```diff
@@ -295,11 +281,283 @@ private static MatrixBlock computeMatrixInverse(Array2DRowRealMatrix in) {
+			throw new DMLRuntimeException("QR2 Decomposition can only be done on a square matrix. "
+				+ "Input matrix is rectangular (rows=" + in.getNumRows() + ", cols=" + in.getNumColumns() + ")");
+		}
+
+		int num_Threads = 1;
+		int m = in.rlen;
+
+		MatrixBlock A_n = new MatrixBlock(m, m, 0.0);
+		A_n.copy(in);
+
+		MatrixBlock Q_n = new MatrixBlock(m, m, 0.0);
+		for(int i = 0; i < m; i++) {
+			Q_n.setValue(i, i, 1.0);
+		}
+
+		ReorgOperator op_t = new ReorgOperator(SwapIndex.getSwapIndexFnObject(), num_Threads);
+		AggregateBinaryOperator op_mul_agg = InstructionUtils.getMatMultOperator(num_Threads);
+		BinaryOperator op_sub = InstructionUtils.parseExtendedBinaryOperator("-");
+		RightScalarOperator op_div_scalar = new RightScalarOperator(Divide.getDivideFnObject(), 1, num_Threads);
+		LeftScalarOperator op_mult_2 = new LeftScalarOperator(Multiply.getMultiplyFnObject(), 2, num_Threads);
+
+		for(int k = 0; k < m; k++) {
+			MatrixBlock z = A_n.slice(k, m - 1, k, k);
+			MatrixBlock uk = new MatrixBlock(m - k, 1, 0.0);
+			uk.copy(z);
+			uk.setValue(0, 0, uk.getValue(0, 0) + Math.signum(z.getValue(0, 0)) * Math.sqrt(z.sumSq()));
+			op_div_scalar = (RightScalarOperator) op_div_scalar.setConstant(Math.sqrt(uk.sumSq()));
+			uk = uk.scalarOperations(op_div_scalar, new MatrixBlock());
+
+			MatrixBlock vk = new MatrixBlock(m, 1, 0.0);
+			vk.copy(k, m - 1, 0, 0, uk, true);
+
+			MatrixBlock vkt = vk.reorgOperations(op_t, new MatrixBlock(), 0, 0, m);
+			MatrixBlock vkvkt = vk.aggregateBinaryOperations(vk, vkt, op_mul_agg);
+			MatrixBlock vkvkt2 = vkvkt.scalarOperations(op_mult_2, new MatrixBlock());
+
+			A_n = A_n.binaryOperations(op_sub, A_n.aggregateBinaryOperations(vkvkt2, A_n, op_mul_agg));
+			Q_n = Q_n.binaryOperations(op_sub, Q_n.aggregateBinaryOperations(Q_n, vkvkt2, op_mul_agg));
+		}
+		// Q_n= Q
+		// A_n = R
+		return new MatrixBlock[] {Q_n, A_n};
+	}
+
+	/**
+	 * Function that computes the Eigen Decomposition using the QR algorithm.
+	 * Caution: check if the QR algorithm is converged, if not increase iterations
+	 * Caution: if the input matrix has complex eigenvalues results will be incorrect
+	 *
+	 * @param in Input matrix
+	 * @return array of matrix blocks
+	 */
+	private static MatrixBlock[] computeEigenQR(MatrixBlock in) {
+		return computeEigenQR(in, 300);
+	}
+
+	private static MatrixBlock[] computeEigenQR(MatrixBlock in, int num_iterations) {
+		if(in.getNumRows() != in.getNumColumns()) {
+			throw new DMLRuntimeException("Eigen Decomposition (QR) can only be done on a square matrix. "
+				+ "Input matrix is rectangular (rows=" + in.getNumRows() + ", cols=" + in.getNumColumns() + ")");
+		}
+		int num_Threads = 1;
+		double tol = 1e-10;
+		int m = in.rlen;
+
+		AggregateBinaryOperator op_mul_agg = InstructionUtils.getMatMultOperator(num_Threads);
+
+		MatrixBlock Q_prod = new MatrixBlock(m, m, 0.0);
+		for(int i = 0; i < m; i++) {
+			Q_prod.setValue(i, i, 1.0);
+		}
+
+		for(int i = 0; i < num_iterations; i++) {
```

Comment:
add todo or make this loop parallel.

Reply:
We are not sure if we can parallelize that loop, since it is a iterative algorithm.
Or is there another way of making it parallel?

Reply:
unclear comment from my side again,
i was referring to using the num_threads argument and make the instructions parallel.
this would entail adding the num_threads to the computeQR2 call.